### DISCUSSION TITLE
Alternative Proposal for Reducing Allocations when using Collection Initializer Syntax

### DISCUSSION THREAD

**Discussion by Saalvage** [dotnet/csharplang#6186] [Language Ideas]
There have been numerous proposals trying to incorporate pre-allocation with the collection initializer syntactic sugar. All of them (well, all of them that I've managed to find) try to accomplish this by doing some weird and opaque modifications to the constructor being called, which results in numerous issues, which are outlined elsewhere.

The alternative that I would like to propose is the following:
If the type on which the collection initializer syntax is used has a method of the signature `EnsureCapacity(int)`, a call to this method is synthesized with the number of elements within the collection initializer's braces as the sole argument. It should occur after the constructor call and before any of the `Add` calls are made.

Example:
```cs
// syntactic sugar
var list1 = new List() { 1, 2, 3 };

// current translation:
var list2 = new List();
list2.Add(1);
list2.Add(2);
list2.Add(3);

// newly proposed translation:
var list3 = new List();
list3.EnsureCapacity(3);
list3.Add(1);
list3.Add(2);
list3.Add(3);
```

This has numerous benefits:
- It would not require any library modifications. All standard classes that would be able to profit from this are already equipped with the method in question.
- It works very well with custom types. If a type follows the design laid out by the standard collection types closely there is a high probability that it would be able to benefit from this change without modification.
- The chance for error is incredibly minor. The capacity of a collection is inherently an implementation detail and even then, assuming `EnsureCapacity` follows the same allocation strategy as `Add` does on a capacity overrun there would be no observable change of behavior (except the speedup). Also, as opposed to a constructor taking an `int`, a method named `EnsureCapacity` very clearly communicates its semantics. So unless `EnsureCapacity` does something that is very different from what you'd expect from a method called `EnsureCapacity` there would also be no observable change of behavior for custom types.
- The change is non-destructive. `EnsureCapacity` is not designed to reduce the capacity, only increase it, meaning that if the user has more information about the final size of the collection the behavior of the pre-allocating constructor would not be overriden if it is used.

The only real downside as compared to the proposals suggesting the use of the pre-allocating constructor is the likely necessity of two allocations in most cases, rather than one. Assuming that the parameterless constructor allocates and the collection initializer's capacity is beyond its initial capacity it has to allocate again. But O(1) allocations are still immensely better than O(log n) allocations and combined with the ease of implementation and the non-destructive nature of the change I believe that this is something that can speed up things with practically no downsides!

### Amendment:
In order to prevent any potential issues from occurring due to this change the likely best approach is to individually toggle this feature for containers that have a fitting `EnsureCapacity` method via an attribute. This means that all standard containers, as well as all custom container types that wish to support this behavior can do so, while everything else remains as is, and therefore unbroken!

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/5354

**Reply by Saalvage** [reply]
That's a new feature, while this would only augment an existing one. Also that feature lacks support for `Dictionary`, while this one supports it.

**Reply by HaloFour** [reply]
> That's a new feature, while this would only augment an existing one.

IMO that makes this proposal more of a liability, it will impact and change the behavior of a massive amount of existing code.  That could lead to subtle breaking changes.  That same issue has stymied considering having collection initializers use `AddRange` also.

I'd personally like to see improvements in this space but the challenge would be how to do that while limiting the potential for breaking changes.

**Reply by Saalvage** [reply]
@HaloFour I think the important distinction is that this would *not* simply alter behavior, as would a change to `AddRange`, but instead amend it with a single additional function call in a specific set of circumstances. All existing calls will still take place in the same order as originally, except that they are prefixed with an additional call to a really specific method.

I have a really, really hard time thinking about any potential, even slightly realistic breakage that would be caused by this. Assuming the `EnsureCapacity` methods in the standard library employ the same growing strategy as do the `Add` methods, all outside behavior is exactly identical.

This leaves custom types fulfilling the criteria for the collection initializer syntax, which already constitutes an incredibly small amount of code. And even in that case, what slightly realistic implementation would `EnsureCapacity` need to have in order to cause anything to break?

I understand that this needs to be thought about hard and long, but I just can't come up with anything that isn't just intentionally trying to be a breaking change. Which either speaks for my lack of creativity and/or knowledge or the relative safety of my proposal.

Lastly also consider: This does certainly affect a massive amount of code, but in 99% of cases where standard types are used that's 100% for the better! So even if this would only be restricted to built-in types to eradicate any remaining possibility of a breaking change, a massive amount of code could benefit from it!

**Reply by HaloFour** [reply]
> This does certainly affect a massive amount of code, but in 99% of cases where standard types are used that's 100% for the better! 

I'd agree.  It's that 1%, because 1% of "massive amount" is still a _lot_.  It's difficult for the language to assume what `EnsureCapacity` should do, and even if it exists in a standard form on the BCL collections that doesn't mean it exists and works properly universally.  Considering the bar for a type to be able to support collection initializers is quite low that's a lot of custom classes that would now be assumed to work "correctly".

Some of this may be moot with [collection literals](https://github.com/dotnet/csharplang/issues/5354) as it's proposed that when the target type is a collection (e.g. `List<T>`) and the type has a constructor that accepts a capacity parameter that the compiler would synthesize code that would call that constructor automatically:

```cs
List<int> list = [ 1, 2, 3 ];

// translated as
List<int> $temp = new List<int>(capacity: 3);
$temp.Add(1);
$temp.Add(2);
$temp.Add(3);
List<int> list = $temp;
```

**Reply by Saalvage** [reply]
@HaloFour To be honest, I feel like `EnsureCapacity` is pretty verbosely named. What else should a method called `EnsureCapacity` on a type implementing `IEnumerable<T>` and having an `Add(T)` method do? I think I might need to reiterate that the bar for this feature would be slightly higher than the one for collection initializers. If the `EnsureCapacity` method doesn't exist it's not called, existing types without the method would continue working correctly.

I disagree with the fact that collection literals would make this moot. Sure, it's a feature that would supersede this one, but on one hand, (and I'm saying this as someone who knows very little about the design process, so I could be very wrong on this, but) it does appear that they have quite some more way to go, while this proposal is very straightforward without pretty much any complexity. On the other hand that is a new feature, while this would help improve a ton of existing code!

I don't have knowledge on how to gain this data, but I'm sure someone does. I really highly doubt that you'll find any piece of code in the wild for which calling `EnsureCapacity` on a custom type in the manner described here would be a breaking change in any shape way or form. I'd genuinely be interested in knowing.

Heck, if having it apply to custom types really is the only concern, then just ditch that and only have it apply to the built-in ones. If you want have it check an attribute on the `EnsureCapacity` method instead!

If it really comes down to that I am of the opinion that improving 99% of code instead of 100% is still so much more preferable to improving 0%.

**Reply by CyrusNajmabadi** [reply]
> If it really comes down to that I am of the opinion that improving 99% of code instead of 100% is still so much more preferable to improving 0%.

The danger is potentially breaking existing code. Either through functional breaks, or through behavioral ones. 

For example, say we set the capacity low due to only seeing one (or zero) values in the initializer. Then later the code adds many more values. It's possible that there may be implementations out there that perform worse if we now insert such a call insitu.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

I was kinda thinking that collection literals have a similar issue.  My example above would set the initial capacity to 3, but the default for `List<T>` is 4, so if you add one more item to that list the performance would be worse with collection literals.  And I can imagine that an analyzer might offer a "fixit" to convert between simple collection initializers to collection literals.  Maybe that's where some smarts should be applied to take into consideration the number of elements in the initializer.  It's a tricky business.

**Reply by Saalvage** [reply]
@CyrusNajmabadi I understand the issue of breaking changes. Yet I still do not see how that applies to this in any sensible way. `EnsureCapacity` does not *set* a capacity, it *ensures* one, meaning that the capacity afterwards will be >= than the requested one.

How would two different growing algorithms have to be implemented to show the described performance degradation? In my opinion the implementation would have to be really rubbish to exhibit such behavior, and at that point I don't think the language design should care.

P.S.: At least `List` even specifies the growing algorithm for `EnsureCapacity`. It's duplicating the capacity until it's >=, so pretty much as optimal as you can get. I have to imagine the other containers behave similarly, making that point moot for using `EnsureCapacity`. Although I do see how that could be a genuine issue for using the pre-allocating constructor as that one at least claims to genuinely *set* the capacity.

**Reply by HaloFour** [reply]
@Saalvage 

> In my opinion the implementation would have to be really rubbish to exhibit such behavior, and at that point I don't think the language design should care.

Considering this would apply to every implementation, rubbish or otherwise, it is necessary for the language design to care.  If this were a new feature it'd be different, as it's expected that an implementation would be expected to adhere to a specific set of guidelines.  But this is a feature that applies to any collection-like type written in the past ~15 years.

IMO, wading into this territory at all I think requires a new pattern to which supporting collection types would be expecting to adhere, and not accidentally.  For example, a new family of attributes that the compiler would recognize that would enable better code generation in these scenarios.  However, if we're going that route I think we'd be better off having a completely different design to collection initializers which would dispense with the rest of the overhead, support immutable collection types, etc.

**Reply by Saalvage** [reply]
@HaloFour I've already forfeited the automatic application to all types in order to just support the 99% of code that only uses standard containers, which should definitely fulfill what is required of them here.

Yeah I've already proposed the attribute solution. Just provide an attribute that applies to the `EnsureCapacity` method and you're good to go. All code will continue to function and 99% of code will be able to enjoy the performance gain and if the remaining 1% wants to enjoy it as well all they have to do is apply a single attribute. That sounds like a sweet deal to me!

Sure, introduce cool new features, but when you have the chance to improve a ton of preexisting code with no downside whatsoever, you should take it as well! (That's what I think at least ^^)

**Reply by CyrusNajmabadi** [reply]
Basically, it's a tradeoff on risk/reward. Tbh, there feels like there's lots reward, and potential risk on patterns that have genuinely been around nearly 20 years. I'm not really sure there would be appetite for that. But we can see how the community feels about this PR, and if anyone on the ldm thinks this is worthwhile.

**Reply by CyrusNajmabadi** [reply]
>  ensures one, meaning that the capacity afterwards will be >= than the requested one.

This seems like this proposal will be matching on shape, not on any specified API. So it can't really be said what the behavior will be. 

Here's a concrete example:
Someone provides a collection, and they have this method, but never implement it (because they never got around to it, and none of their code calls it). They upgrade, and now their code starts throwing.

**Reply by Saalvage** [reply]
@CyrusNajmabadi Could you elaborate on the risk that you see if this were to be applied to standard containers exclusively? Keep in mind that it would require a recompilation, so if the old code were to update it would also have to use the new standard containers, which we'd expect to adhere to the quality standards outlined above.

Is it just the fear that there could be something that everyone just fails to consider (which would be pretty fair)? Or some actually existing change that you could imagine?

**Reply by CyrusNajmabadi** [reply]
> with no downside whatsoever,

Risk is a downside. It also occupies a lot of time for the team. So features really need to be substantively beneficial and balance the pros/cons out in order to be taken.

**Reply by CyrusNajmabadi** [reply]
> Could you elaborate on the risk that you see if this were to be applied to standard containers exclusively? 

I'm not sure how we would define "a standard container". Primarily, my concern then would be us doing that, and then having other groups coming in (including other MS teams) saying: why are we not getting this new behavior for our types :-)

**Reply by CyrusNajmabadi** [reply]
If this were done with an attribute, or some other way to opt in, it would likely be very low risk. My genuine feeling though would be around asking if this was worth it. 

After all, there will still be an extra allocation here. First at construction time, then when ensuring the capacity.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> If this were done with an attribute, or some other way to opt in, it would likely be very low risk. My genuine feeling though would be around asking if this was worth it.

That's kinda why I think it's more worthwhile to consider a new approach to collection initializers in general.  Perhaps a factory method that accepts a `ROS<T>` or the like, eliminating the need for more allocations and multiple virtual calls.  It's more work, but that seems more palatable to adding what would likely be a family of separate attributes to enable patterns piecemeal.

**Reply by Saalvage** [reply]
I still think the cost of the virtual method calls are likely dwarfed by the need for O(log n) reallocations, especially when it concerns very large lists being initialized directly.

So, to recap a bit: the attribute approach is the one with by far the most merit, basically no risk, but quite some reward.

I honestly completely understand if this is not seen as beneficial enough to consider for implementation. ^^
Thank all of you for the active discussion!

**Reply by CyrusNajmabadi** [reply]
You're welcome :-)

**Reply by CyrusNajmabadi** [reply]
Ultimately @Saalvage I think the vast majority of usages don't really care. They're not going to hit these capacities, or they won't care about some small number of gen-1 allocs. 

So it's only the case if a hotspot *and* a ton of elements. But in that case, I would expect someone can just profile, see an issue and just manually set capacity. Yes, it's more onerous, but it's the very rare case. 

The reason I'm focusing on the collection literal improvements is so that we can tackle a lot of issues at once, making the design and engineering effort worthwhile :-)

**Reply by Saalvage** [reply]
@CyrusNajmabadi That's very much true. I wish you the best of luck with collection literals!!
